src/render/line/draw.rs

```diff
@@ -52,42 +52,52 @@ fn messages(
         tokens.clear();
         let blocks_drawn_during_previous_tick = state.blocks_per_line.pop_front().unwrap_or(0);
-        let message_block_len = origin.width();
-        current_maximum = current_maximum.max(message_block_len);
-        if state.message_origin_size.len() == max_height {
-            state.message_origin_size.pop_front();
-        }
-        state.message_origin_size.push_back(message_block_len);
 
-        let color = to_color(*level);
-        tokens.push(" ".into());
-        if timestamp {
-            tokens.push(
-                brush
-                    .style(color.dimmed().on(Color::Yellow))
-                    .paint(crate::time::format_time_for_messages(*time)),
-            );
-            tokens.push(Style::default().paint(" "));
-        } else {
-            tokens.push("".into());
+        match envelope {
+            Envelope::RawMessage(content) => {
+                // RawMessages have already been split into single lines without newlines appended,
+                // and so can be rendered directly as ANSIStrings.
+                tokens.push(ANSIString::from(content))
+            }
+            Envelope::Message(Message {
+                time,
+                level,
+                origin,
+                message,
+            }) => {
+                let message_block_len = origin.width();
+                current_maximum = current_maximum.max(message_block_len);
+                if state.message_origin_size.len() == max_height {
+                    state.message_origin_size.pop_front();
+                }
+                state.message_origin_size.push_back(message_block_len);
+
+                let color = to_color(*level);
+                tokens.push(" ".into());
+                if timestamp {
+                    tokens.push(
+                        brush
+                            .style(color.dimmed().on(Color::Yellow))
+                            .paint(crate::time::format_time_for_messages(*time)),
+                    );
+                    tokens.push(Style::default().paint(" "));
+                } else {
+                    tokens.push("".into());
+                };
+                tokens.push(brush.style(Style::default().dimmed()).paint(format!(
+                    "{:>fill_size$}{}",
+                    "",
+                    origin,
+                    fill_size = current_maximum - message_block_len,
+                )));
+                tokens.push(" ".into());
+                tokens.push(brush.style(color.bold()).paint(message));
+            }
         };
-        tokens.push(brush.style(Style::default().dimmed()).paint(format!(
-            "{:>fill_size$}{}",
-            "",
-            origin,
-            fill_size = current_maximum - message_block_len,
-        )));
-        tokens.push(" ".into());
-        tokens.push(brush.style(color.bold()).paint(message));
+
         let message_block_count = block_count_sans_ansi_codes(&tokens);
```

Comment:
Hm: it looks like `unicode-width` does not ignore ANSI control characters, so raw messages containing ANSI characters do not fully overdraw. See https://github.com/unicode-rs/unicode-width/issues/24.